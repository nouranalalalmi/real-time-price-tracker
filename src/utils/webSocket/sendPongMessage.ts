export const sendPongMessage = (id: number, sendMessage: Function) => {
  const message = JSON.stringify({
    method: 'pong',
    params: [],
    id,
  });

  sendMessage(message);
};
