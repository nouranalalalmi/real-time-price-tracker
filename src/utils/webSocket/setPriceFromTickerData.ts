export const setPriceFromTickerData = (data: any, setPrices: Function) => {
  const { s: symbol, p: priceChange, P: priceChangePercent, c: lastPrice } = data;

  setPrices({
    [symbol]: {
      priceChange,
      priceChangePercent,
      lastPrice,
    },
  });
};
