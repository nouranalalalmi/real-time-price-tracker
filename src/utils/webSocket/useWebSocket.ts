import { useEffect } from 'react';

import { COINS } from '@/constants';
import { usePriceStore } from '@/stores/pricesStore';

export const useWebSocket = () => {
  const setPrices = usePriceStore(state => state.setPrices);

  useEffect(() => {
    const socket = new WebSocket('wss://stream.binance.com:443/ws');

    socket.onopen = () => {
      socket.send(
        JSON.stringify({
          method: 'SUBSCRIBE',
          params: COINS.map(coin => `${coin.tickerId}@ticker`),
          id: 1,
        })
      );
    };

    socket.onmessage = event => {
      const data = JSON.parse(event.data);

      if (data.event === 'ping') {
        console.log('a ping happened', data);

        socket.send(
          JSON.stringify({
            method: 'pong',
            params: [],
            id: data.id,
          })
        );
      } else {
        const { s: symbol, p: priceChange, P: priceChangePercent, c: lastPrice } = data;
        setPrices({
          [symbol]: {
            priceChange,
            priceChangePercent,
            lastPrice,
          },
        });
      }
    };

    socket.onerror = error => {
      console.error('WebSocket error', error);
    };

    // Clean up the WebSocket connection when the component unmounts
    return () => {
      console.log('close connection');

      socket.close();
    };
  }, [setPrices]);
};
