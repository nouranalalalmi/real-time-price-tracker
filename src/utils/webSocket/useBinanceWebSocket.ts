import { useState } from 'react';
import useWebSocket from 'react-use-websocket';

import { CONNECTION_STATUS, COINS } from '@/constants';
import { usePriceStore } from '@/stores/pricesStore';

import { sendPongMessage } from './sendPongMessage';
import { setPriceFromTickerData } from './setPriceFromTickerData';

export const useBinanceWebSocket = () => {
  const [error, setError] = useState<null | string>(null);
  const setPrices = usePriceStore(state => state.setPrices);

  const { sendMessage, readyState } = useWebSocket('wss://stream.binance.com:443/ws', {
    shouldReconnect: () => true,
    onError: error => {
      setError(`Failed to establish connection: ${error.type}`);
    },
    onOpen: () => {
      setError(null);
      sendMessage(
        JSON.stringify({
          method: 'SUBSCRIBE',
          params: COINS.map(coin => `${coin.tickerId}@ticker`),
          id: 1,
        })
      );
    },
    onMessage: message => {
      const data = JSON.parse(message.data);

      if (data.e === 'ping') {
        sendPongMessage(data.id, sendMessage);
      } else if (data.e === '24hrTicker') {
        setPriceFromTickerData(data, setPrices);
      }
    },
  });

  const connectionStatus = CONNECTION_STATUS[readyState];

  return { connectionStatus, error };
};
