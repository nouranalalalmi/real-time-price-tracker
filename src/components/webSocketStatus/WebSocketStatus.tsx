'use client';
import { useBinanceWebSocket } from '@/utils/webSocket/useBinanceWebSocket';

export const WebSocketStatus = () => {
  const { connectionStatus, error } = useBinanceWebSocket();

  return (
    <div className="border bg-white p-1 text-center text-xs shadow-sm">
      <p className=" font-semibold">Web socket status: {connectionStatus}</p>
      {error && <p className="text-red-600">{error}</p>}
    </div>
  );
};
