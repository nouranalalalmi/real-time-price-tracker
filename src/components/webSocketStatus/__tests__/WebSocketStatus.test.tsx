import { render, screen } from '@testing-library/react';

import { useBinanceWebSocket } from '@/utils/webSocket/useBinanceWebSocket';

import { WebSocketStatus } from '../WebSocketStatus';

jest.mock('@/utils/webSocket/useBinanceWebSocket');

describe('WebSocketStatus', () => {
  it('should display connection status', () => {
    (useBinanceWebSocket as jest.Mock).mockReturnValue({
      connectionStatus: 'connected',
      error: null,
    });

    const { container } = render(<WebSocketStatus />);

    expect(screen.getByText('Web socket status: connected')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should display error message', () => {
    (useBinanceWebSocket as jest.Mock).mockReturnValue({
      connectionStatus: 'disconnected',
      error: 'Test error',
    });

    const { container } = render(<WebSocketStatus />);

    expect(screen.getByText('Web socket status: disconnected')).toBeInTheDocument();
    expect(screen.getByText('Test error')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
