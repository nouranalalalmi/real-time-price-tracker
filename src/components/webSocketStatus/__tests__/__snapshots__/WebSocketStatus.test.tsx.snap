// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebSocketStatus should display connection status 1`] = `
<div>
  <div
    class="border bg-white p-1 text-center text-xs shadow-sm"
  >
    <p
      class=" font-semibold"
    >
      Web socket status: 
      connected
    </p>
  </div>
</div>
`;

exports[`WebSocketStatus should display error message 1`] = `
<div>
  <div
    class="border bg-white p-1 text-center text-xs shadow-sm"
  >
    <p
      class=" font-semibold"
    >
      Web socket status: 
      disconnected
    </p>
    <p
      class="text-red-600"
    >
      Test error
    </p>
  </div>
</div>
`;
